File: java8/src/main/java/io/github/ocirne/rosetta/ingredientsmerger/IngredientsMerger.java
package io.github.ocirne.rosetta.ingredientsmerger;

import io.github.ocirne.rosetta.ingredientsmerger.given.Ingredient;

import java.util.List;

interface IngredientsMerger {

    List<Ingredient> merge(List<Ingredient> list1, List<Ingredient> list2);
}

File: kotlin/src/main/kotlin/io/github/ocirne/rosetta/ingredientsmerger/IngredientsMerger.kt
package io.github.ocirne.rosetta.ingredientsmerger

import io.github.ocirne.rosetta.ingredientsmerger.given.Ingredient

object IngredientsMerger {

    fun merge(list1: List<Ingredient>, list2: List<Ingredient>): List<Ingredient> {
        return (list1 + list2)
            .groupingBy { i -> i.name }
            .fold(0) { total, ingredient -> total + ingredient.amount }
            .map { (name, amount) -> Ingredient(name, amount) }
    }
}
